packages/react-next/src/components/Pivot/useRefEffect.ts

```diff
@@ -0,0 +1,45 @@
+import * as React from 'react';
+
+/**
+ * Creates a ref, and calls a callback whenever the ref changes to a non-null value.
+ * The callback can optionally return a cleanup function that'll be called before the value changes or is set to null.
+ *
+ * This can be used to work around a limitation that useEffect cannot depend on `ref.current` (see
+ * https://github.com/facebook/react/issues/14387#issuecomment-503616820).
+ *
+ * @param callback - Called whenever the ref's value changes to non-null. Can optionally return a cleanup function.
+ * @param initialValue - (Optional) The initial value for the ref.
+ *
+ * @returns A tuple of `[ref, setRef]`:
+ * * `ref` - Read-only view of the ref's current value
+ * * `setRef` - A ref setter function that can be used to change the ref's current value.
+ */
+export function useRefEffect<T>(
+  callback: (value: T) => (() => void) | void,
+  initialValue: T | null = null,
+): [/*ref:*/ React.RefObject<T>, /*setRef:*/ (value: T | null) => void] {
+  type RefData = {
+    readonly ref: React.MutableRefObject<T | null>;
+    readonly setRef: (value: T | null) => void;
+    currentCallback: (value: T) => (() => void) | void;
+    cleanup?: (() => void) | void;
+  };
+
+  const data = React.useRef<RefData>({
+    ref: { current: initialValue },
+    setRef: value => {
+      if (data.ref.current !== value) {
+        if (data.cleanup) {
+          data.cleanup();
+        }
+        data.ref.current = value;
+        data.cleanup = value !== null ? data.currentCallback(value) : undefined;

```

Comment:
shouldn't we still call the callback when `value` is set to null ?

Reply:
The cleanup function can be used to do whatever you need to do when the ref is set to null.

React sets refs to `null` when the component is unmounted. That's how useRefEffect guarantees that `cleanup` will always run on unmount. If I call the callback when it's set to `null`, then the callback could return another `cleanup` function, which would never be run.

Thinking about it more, I could just run `cleanup` immediately after calling `callback` if the ref is null. Seems like potentially confusing behavior though. Thoughts?

Reply:
Here is how i think (correct me if wrong) - the behavior here should be consistent with how `useEffect` behaves. for `useEffect`, every time deps changes, cleanup will be called and callback will be called. I'm curious when `useEffect` calls the final `cleanup`.

Reply:
ic. `useEffect` calls the final clean up when component is unmounted. see: https://codesandbox.io/s/useeffect-cleanup-qjdmb

Reply:
React sets refs to null when the component is unmounted. So the current behavior *is* aligned with `useEffect` that has a dependency on `[ref.current]`:
* The first time the ref is set to non-null, `callback` is called.
* When the ref's value changes (non-null), `cleanup` is called, then `callback` is called.
* When the ref's value changes to null (i.e. on unmount), then only `cleanup` is called.

See: https://codesandbox.io/s/misty-flower-sfbno?file=/src/App.tsx